inception/inception-api-annotation-experimental/src/main/java/de/tudarmstadt/ukp/inception/experimental/api/AnnotationSystemAPIImpl.java

```diff
@@ -17,402 +17,796 @@
+            // TODO server decide on next document
+            if (aNewDocumentRequest.getDocumentId() == 0) {
+                aNewDocumentRequest.setDocumentId(41714);
             }
 
-            message.setViewportText(getViewportText(aClientMessage, cas));
-
-            message.setAnnotations(
-                    filterAnnotations(getAnnotations(cas, aClientMessage.getProject()),
-                            aClientMessage.getViewport()));
+            CAS cas = getCasForDocument(aNewDocumentRequest.getUserName(),
+                    aNewDocumentRequest.getProjectId(), aNewDocumentRequest.getDocumentId());
 
-            annotationProcessAPI.handleSendDocumentRequest(message, aClientMessage.getClientName());
+            if (aNewDocumentRequest.getViewport().getDisabledLayers() == null) {
+                aNewDocumentRequest.getViewport().setDisabledLayers(new ArrayList<>());
+            }
+            Pair<List<Span>, List<Relation>> annotations = getAnnotations(cas,
+                    aNewDocumentRequest.getProjectId(),
+                    getValidLayers(aNewDocumentRequest.getProjectId(),
+                            aNewDocumentRequest.getViewport().getDisabledLayers()));
+
+            List<Span> spans = filterSpans(annotations.getKey(), aNewDocumentRequest.getViewport());
+
+            List<Relation> relations = filterRelations(annotations.getValue(), cas,
+                    aNewDocumentRequest.getViewport());
+
+            NewDocumentResponse message = new NewDocumentResponse(
+                    toIntExact(documentService.getSourceDocument(aNewDocumentRequest.getProjectId(),
+                            aNewDocumentRequest.getDocumentId()).getId()),
+                    getViewportText(cas, aNewDocumentRequest.getViewport()), spans, relations);
+            annotationProcessAPI.sendNewDocumentResponse(message,
+                    aNewDocumentRequest.getClientName());
         }
         catch (Exception e) {
+            LOG.error("Exception has been thrown during handleNewDocument()");
             e.printStackTrace();
-            createErrorMessage(e.getMessage(), aClientMessage.getClientName());
+            createErrorMessage(e.getMessage(), aNewDocumentRequest.getClientName());
         }
 
     }
 
     @Override
-    public void handleViewport(ClientMessage aClientMessage) throws IOException
+    public void handleNewViewport(NewViewportRequest aNewViewportRequest) throws IOException
     {
-        CAS cas = getCasForDocument(aClientMessage.getUsername(), aClientMessage.getProject(),
-                aClientMessage.getDocument());
+        try {
+            CAS cas = getCasForDocument(aNewViewportRequest.getUserName(),
+                    aNewViewportRequest.getProjectId(), aNewViewportRequest.getDocumentId());
+
+            if (aNewViewportRequest.getViewport().getDisabledLayers() == null) {
+                aNewViewportRequest.getViewport().setDisabledLayers(new ArrayList<>());
```

Comment:
You could change setDisabledLayers(args) to copy args to a final internal list in the Viewport - or let Viewport return e.g. `Collections.emptyList()` if the disabledlayers is null.

Reply:
Thats a lot better, thank you for the hint!